hubble: Metrics server cell

devodev opened the pull request:
### Description

This PR extracts the initialization logic for the Hubble metrics server and the flow processor into its own Hive cell.
Since this work required abstracting the mechanism for retrieving certificates for the Hubble metrics server using `certloader.WatchedServerConfig`, I also re-used it for the Hubble server certificates.

Two notable behaviour changes:
- The metrics server no longer blocks Hubble from starting while waiting for TLS certificates to become available.
- The metrics server now reports warnings/errors in cilium status.
- The metrics server and certloader tls cert watcher now reports their job status in the health table.

Additionally, the default value for the `--hubble-disable-tls` flag is now `true`. The value in helm options is unchanged. This is to prevent errors in hive tests returned by the certloader dependency requiring a filepath set when enabled at hive startup time. Before we did not face issues because the TLS config watcher was always setup in the launch method of Hubble, after the cell has been started. The helm chart always sets the value explicitly, therefore there is no behaviour change caused by this.

See individual commits for details.

Related: #34501
cc @kaworu 

```release-note
Extract the Hubble metrics server into its own Hive cell
Include more details about the Hubble metrics server state in cilium-status
```

### Richer cilium status reporting

We now record and report warnings and errors when starting/running the metrics server.
Previously, we would always report OK even if the server crashed at some point. Additionally, now that Hubble no longer blocks at initialization waiting for the metrics server TLS certificates to become ready, we can report the state of Hubble and the metrics server independently.

Example:
```
$ cilium status
...
Hubble:                  Ok              Current/Max Flows: 4095/4095 (100.00%), Flows/s: 31.79   Metrics: Warning (Waiting for TLS certificates to become available)
```

We could think of expending this in the future to report more information, such as how many metrics handlers are enabled, and even which ones.

Additionally, by using one-shot jobs to register long-lived processes such as the metrics server and certloader tls watchers, we get entries in the health table:
```
$ cilium status --all-health
...
├── hubble
│   ├── hubble-metrics
│   │   ├── job-certloader-tls          [OK] Running (8m33s, x1)
│   │   └── job-hubble-metrics-server   [OK] Running (8m33s, x1)
│   ├── job-certloader-tls              [STOPPED] Running (8m33s, x1)
│   └── job-hubble                      [STOPPED] Running (8m33s, x1)
```

### Certloader cell provider

While working on this, I noticed that the process of creating a FutureWatchedServerConfig and waiting for TLS certificates to become ready was duplicated for both the Hubble metrics server and the observer server. I took this opportunity to refactor that logic into a new `certloader.NewWatchedServerConfigPromise` constructor that can be used with `cell.Provide` and `cell.ProvidePrivate`.

At first, I considered adding a new "reusable" cell, similar to how the pprof cell has been refactored lately and allows providing custom flags for it. The difficulty here is in avoiding collisions/replacement with the provided type, here a promise over a WatchedServerConfig, when the cell is being used multiple times in the same hive. By only providing the constructor, its possible for a cell author to wrap the return type and provide that as a dependency. This is how we are doing it here with both the Hubble metrics server and the Hubble server cells.

devodev committed:
hubble: Add static metrics flow processor type

Abstract the OnDecodedFlow hook that we register when using
static metrics in a StaticFlowProcessor type to align with
the existing DynamicFlowProcessor.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
crypto/certloader: Add reusable cell provider for WatchedServerConfig

Add a new constructor that can be used in a cell provide call that
provides a Promise over a WatchedServerConfig. The promise resolves
when the certificate files become available. When disabled, the
promise will be nil, allowing to detect whether certloader was enabled
or not. On hive stop, the cell will take care of stopping the config
watcher if it exists.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
hubble: Add metrics FlowProcessor interface

Abstract the static and dynamic flow processors behind an interface.
This will allow providing the interface to the Hubble cell and have
hubble register the OnDecodedFlow hooks.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
hubble: Integrate hubble-metrics cell in Hubble cell

Replace the current initialization logic for the metrics server
and flow processor with the new hubble-metrics cell.

Update the hubble status reporting to now use the new metrics server
status.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
api/healthz: Introduce HubbleMetricsStatus object

This new status object will be used by the incoming hubble-metrics
cell to report its status independently from the Hubble cell.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
hubble: Introduce hubble-metrics cell

Create a new Hive cell that provides the hubble metrics server
and the metrics flow processor which is meant to be hooked in
the hubble observer to perform metrics recording.

The metrics server also implements a new Server interface that
will be used to report the server state from cilium-status.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev committed:
hubble: Report hubble-metrics server status in cilium-status

Upgrade the status reporting for the metrics server used in cilium-status
and include warnings and errors observed when starting/running the server.

More specifically, record and report a warning status when the server is waiting
for the TLS certificates to become ready, report an error if we failed to wait for
certificates and report an error if we failed serving http metrics.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

glrf commented on pkg/crypto/certloader/cell.go:
```diff
@@ -0,0 +1,107 @@
+// a WatchedServerConfig. The config can be used to obtain a tls.Config which can
+// transparently reload certificates between two connections. The promise resolves
+// when the certificates become ready and have been loaded.
+func Cell[Cfg cell.Flagger](cfg Cfg) cell.Cell {
```
I'm not a hive/cell expert, so I might be wrong/missing the point, but:

1. Does this actually allow us to instantiate the certloader twice? We'd then have two cells that provide `promise.Promise[*WatchedServerConfig]`, right? For the pprof cell this isn't a problem, because it's only instantiated once per hive, but for the certloader usecase we'd want to have multiple certloaders per hive (e.g. hubble metrics, hubble server), don't we?
2. I really struggled for a while until I understood how this solution with two different configs works. I know there is precedence for this, so I'm not going to block this if we keep it, but I'm not a huge fan.

devodev replied:
1. Yes that is the goal ultimately although you raise a good point and I thought I had dealt with that. So what I had in mind is using cell.ProvidePrivate to ensure that the dependency is only visible to the scope where its needed. I realize now that the cell uses Provide. Let me try to change that and see if this all works as expected.  I should probably try to abstract the hubble server certloader and see if it all works, since this work will need to happen as well anyway. Thanks for catching that!
2. I also did need the pprof cell to figure out all that was needed, not extra clear. We discussed with @kaworu and took the leap to try and have a clean abstraction, but it does make it somewhat hard to comprehend, I agree. I guess if this is an area of contention, we could instead revert to using custom certloader cells for hubble metrics and server.

devodev replied:
I should also mention that it would be great if the certloader cell could return an arbitrary typed promise that contains the watchedServerConfig so that you can use multiple ones in the same cell and avoid providePrivate tricks. Maybe there is something to explore similar to how the pprof cell handles a custom config as input 🤔

glrf replied:
Maybe you could actually take a typed config, that is a flagger and can provide the certloader config, and then somehow base the promise on that type.


Veery roughly like this:

```
func Cell[Cfg certloaderConfiger](cfg Cfg) cell.Cell {
	return cell.Module(
		"certloader",
		"Server TLS certificates loader",

		cell.Config(cfg),
		cell.Provide(newWatchedConfig[Cfg]),
	)
}

type certloaderConfiger interface {
	cell.Flagger
	Config() Config
}
```

And then meterics can just call it as

```
	certloader.Cell(defaultCertloaderConfig),
```

Without having to provide the config separately, because `newWatchedConfig` already depends on the provided type

devodev replied:
That is a good idea. I like it and tested it successfully.

For the part about declaring the cell twice, I tested the ProvidePrivate change and it does not work because its in a module. I'll have to rework that part to ensure it works. It might be easier to manually create the cell + config and put the core logic in the certloader package for easy re-use. Putting in draft for now to experiment a bit.

devodev replied:
After much trial and error, I reworked the PR and the logic around providing the certloader WatchServerConfig.

I've updated the PR description with the details.

glrf commented on pkg/hubble/metrics/cell/cell.go:
```diff
@@ -0,0 +1,158 @@
+
+	// metrics flow processor
+	var fp metrics.FlowProcessor
+	switch {
```
Nit: We have a slight behavior change here. Before, if both static and dynamic Hubble metrics where enabled, we returned an error, now we just silently start the static metrics. Could we at least log something?

devodev replied:
Thats not the case, I made sure to use ValidatedConfig, similar to what I did in the exporter, to ensure that this invariant was taken care of by hive. It would not get to this point if a failure occurred from validating the config. Maybe I can add a comment here to make this clearer since its true we use a switch statement deliberately because we know the invariant is handled early.

glrf replied:
Aah. Yes. You're right. I totally missed that

glrf reviewed (commented):
Thanks! Overall this looks like a huge improvement!

devodev committed:
hubble: Replace hubble server certloader logic with cell

Replace the existing logic to setup and wait for certificates to become
available with a new certloader cell. It re-uses the new hive constructor:
certloader.NewWatchedServerConfigPromise and provides a promise
that the Hubble cell can await in its launch function.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

glrf commented on pkg/hubble/metrics/cell/certloader.go:
```diff
@@ -0,0 +1,52 @@
+	cell.ProvidePrivate(certloader.NewWatchedServerConfigPromise),
+	cell.ProvidePrivate(func(p promise.Promise[*certloader.WatchedServerConfig]) tlsConfigPromise {
+		return p
+	}),
```
Nit/Personal preference(feel free to ignore): I'm not a huge fan of `Provide` chains that pass along intermediate values, it always takes me a second to parse. I think you could also change this to the following, which at least to me is a bit more readable and we completely avoid the problem of providing any certloader package types to the dependency graph:

```go
cell.ProvidePrivate(func(lc cell.Lifecycle, log *slog.Logger, cfg certloaderConfig) (tlsConfigPromise, error) {
	 return certloader.NewWatchedServerConfigPromise(lc, log, certloader.Config(cfg))
}),
```

devodev replied:
I think this is reasonable and it does make it more succinct, will update.

glrf reviewed (approved):
LGTM now. Thanks!

chancez commented on pkg/crypto/certloader/cell.go:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	resolver, promise := promise.New[*WatchedServerConfig]()
+	ctx, cancel := context.WithCancel(context.Background())
```
This looks funky to me since your using a `context.Background` here. The `cell.HookContext` seems to provide a context we can use. The cancel gets called in `OnStop`, which I assume is to cancel the polling happening in `OnStart`. Is that necessary if we use the `cell.HookContext`?

devodev replied:
Yes because the cell.HookContext is very short and act as a timeout for the "start" hook itself, not what is being spawned from it. Hence why we use stop to stop that context. Its similar to how other cells manage that.

devodev replied:
I reviewed this section and updated it to move the creation of the watcher inside the OnStart hook which overall makes more sense since it does some setting up such as spawning goroutines and accessing the disk. The cell constructor should do as less as possible.

chancez replied:
Hmmm, yeah I can see most usage of the `OnStart` hook seems to not use, or ignore the context provided. This seems suboptimal, perhaps something in the hive lifecycle hooks is missing. @joamaki is this really how it works?

I'd really like to avoid any usage of `context.Background` as it makes it hard to tell if it's correctly handling cancellation, but apparently we're using it all over already because everyone seems to be ignoring this context.

chancez replied:
I've pinged @joamaki on slack about it. For now this is fine since it seems to follow established patterns.

joamaki replied:
You should use the `job` package instead of creating a context and forking goroutines from start hooks. 

https://pkg.go.dev/github.com/cilium/hive/job
https://docs.cilium.io/en/latest/contributing/development/hive/#job-groups

The `job.Group` type is provided for each module so you can just depend on that instead of `cell.Lifecycle`.
If a job is added _before_ start then it's appended to the lifecycle, if a job is added to the group after starting it'll start immediately.

Here's one pretty minimal example: https://github.com/cilium/cilium/blob/main/pkg/loadbalancer/writer/node_addr_reconciler.go. 

For example here we'd do something like:

```
func NewWatchedServerConfigPromise(jg job.Group, log *slog.Logger, cfg Config) promise.Promise[*WatchedServerConfig] {
    ...
    resolver, promise := promise.New[*WatchedServerConfig]()

    jg.Add(
       job.OneShot(
           "resolve-server",
            func(ctx context.Context, health cell.Health) error {
                 ...
                 select {
                 case <-ctx.Done():
                     return nil
                 case metricsTLSConfig <- metricsTLSConfigChan:
                 }
                 resolver.Resolve(metricsTLSConfig)
                 return  nil
               }))

    return promise
}
```

Another benefit with using jobs is that their health can be inspected with e.g. `db/show health`. The `cell.Health` interface can be used to update their health.

devodev replied:
Yeah I know about jobs and hesitated to use one here actually. I'll work on updating the impl seeing its the preferred way now.

devodev replied:
Updated to use a job :)

chancez commented on pkg/crypto/certloader/cell_test.go:
```diff
@@ -0,0 +1,132 @@
+		t.Fatalf("failed to stop: %s", err)
+	}
+}
+
```
Can we add a test where the we shutdown before it can find a certificate to make sure we don't deadlock on trying to stop the metrics server polling sequence?

devodev replied:
Sure

devodev replied:
I added a new cell shutdown tests, and by doing so, I also noticed a leaked goroutine from `FutureWatcher()` when files never become ready (since we use goleak in cell tests).

Its not too big of a deal when the watcher is tied to the lifecycle of the hive, but can be unexpected in tests and other situations. To fix, I added a new commit that makes `FutureWatcher` take a context and do the necessary cleanup. See commit message for more details.

All in all, that was a good call, thanks!

chancez commented on pkg/hubble/metrics/cell/server.go:
```diff
@@ -0,0 +1,145 @@
+		},
+		tlsConfigPromise: p.TLSConfigPromise,
+		tlsCtx:           ctx,
+		tlsCancel:        cancel,
```
Can we use the hook context instead?

devodev replied:
See reply above. tldr: we can't, it doesn't have the appropriate lifetime.

devodev committed:
crypto/certloader: Make FutureWatcher take a context

In its current form, FutureWatcher spawns a goroutine and sends
the newly created watcher on a channel when the watched files
become ready. Additionally, the created watcher also starts goroutines
for its file watching mechanism. If files never become ready and we
decide to abandon the watch, there is no way to abort the spawned
goroutines, therefore leaking them.

To mitigate, make FutureWatcher take a context and abort both the
outer goroutine and the created watcher+underlying goroutines on
cancel.

Update all certloader tests to detect and fail if any leaked goroutines
are detected using goleak.

Add a new test that ensures the hive is cleanly shutdown and does
not leak goroutines when watched cert files never become ready.

Signed-off-by: Alexandre Barone <abalexandrebarone@gmail.com>

devodev commented on pkg/crypto/certloader/cell.go:
```diff
@@ -0,0 +1,90 @@
+
+	resolver, promise := promise.New[*WatchedServerConfig]()
+
+	jobGroup.Add(job.OneShot("certloader-tls", func(ctx context.Context, _ cell.Health) error {
```
Bringing to your attention this change where instead of manually spawning a goroutine from lifecycle hooks, I used a one-shot job. Bonus points, we get an entry in the health table for it 🥳

devodev commented on pkg/hubble/metrics/cell/server.go:
```diff
@@ -0,0 +1,130 @@
+		tlsConfigPromise: p.TLSConfigPromise,
+	}
+
+	p.JobGroup.Add(job.OneShot("hubble-metrics-server", func(ctx context.Context, _ cell.Health) error {
```
Bringing to your attention this change where instead of manually spawning a goroutine from lifecycle hooks, I used a one-shot job. Bonus points, we get an entry in the health table for it 🥳

devodev reviewed (commented):
Updated the PR based on feedback regarding usage of hive lifecycle hooks vs one-shot jobs for managing long-lived processes.

Bonus, we now get entries for each processes in the health table:
```
├── hubble
│   ├── hubble-metrics
│   │   ├── job-certloader-tls          [OK] Running (8m33s, x1)
│   │   └── job-hubble-metrics-server   [OK] Running (8m33s, x1)
│   ├── job-certloader-tls              [STOPPED] Running (8m33s, x1)
│   └── job-hubble                      [STOPPED] Running (8m33s, x1)
```
